src/BlockTypes/CollectionStockFilter.php

```diff
@@ -0,0 +1,172 @@
+	 */
+	private function get_stock_filter_html( $stock_counts, $attributes ) {
+		$display_style  = $attributes['displayStyle'] ?? 'list';
+		$stock_statuses = wc_get_product_stock_status_options();
+
+		// check the url params to select initial item on page load.
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended -- Nonce verification is not required here.
+		$selected_stock_status = isset( $_GET[ self::STOCK_STATUS_QUERY_VAR ] ) ? sanitize_text_field( wp_unslash( $_GET[ self::STOCK_STATUS_QUERY_VAR ] ) ) : '';
+
+		$list_items = array_map(
+			function( $item ) use ( $stock_statuses ) {
+				return array(
+					'label' => $stock_statuses[ $item['status'] ],
+					'value' => $item['status'],
+				);
+			},
+			$stock_counts
+		);
+
+		$selected_items = array_values(
+			array_filter(
+				$list_items,
+				function( $item ) use ( $selected_stock_status ) {
+						return $item['value'] === $selected_stock_status;
+				}
+			)
+		);
+
+		// Just for the dropdown, we can only select 1 item.
+		$selected_item = $selected_items[0] ?? array(
+			'label' => null,
+			'value' => null,
+		);
+
+		ob_start();
+		?>
+
+		<?php if ( 'list' === $display_style ) : ?>
+			<div class="wc-block-stock-filter style-list">
+				<ul class="wc-block-checkbox-list wc-block-components-checkbox-list wc-block-stock-filter-list">
+					<?php foreach ( $stock_counts as $stock_count ) { ?>
+						<li>
+							<div class="wc-block-components-checkbox wc-block-checkbox-list__checkbox">
+								<label for="<?php echo esc_attr( $stock_count['status'] ); ?>">
+									<input 
+										id="<?php echo esc_attr( $stock_count['status'] ); ?>" 
+										class="wc-block-components-checkbox__input" 
+										type="checkbox" 
+										aria-invalid="false" 
+										data-wc-on--change="actions.filters.updateProducts" 
+										value="<?php echo esc_attr( $stock_count['status'] ); ?>"
+										<?php checked( strpos( $selected_stock_status, $stock_count['status'] ) !== false, 1 ); ?>
+									>
+										<svg class="wc-block-components-checkbox__mark" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 20">
+											<path d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"></path>
+										</svg>
+										<span class="wc-block-components-checkbox__label">
+											<?php echo esc_html( $stock_statuses[ $stock_count['status'] ] ); ?>
+											<?php
+												// translators: %s: number of products.
+												$screen_reader_text = sprintf( _n( '%s product', '%s products', $stock_count['count'], 'woo-gutenberg-products-block' ), number_format_i18n( $stock_count['count'] ) );
+											?>
+											<span class="wc-filter-element-label-list-count">
+												<span aria-hidden="true">
+													<?php echo esc_html( $stock_count['count'] ); ?>
+												</span>
+												<span class="screen-reader-text">
+													<?php esc_html( $screen_reader_text ); ?>
+												</span>
+											</span>
```

Comment:
The count is always displayed even the corresponding inspector setting is set to false.